### PR TITLE
feat(data): Convert the Pile dataset to hf format

### PR DESCRIPTION
**Pull request by boss-chanon**
## Why this PR
make pipeline for convert pile dataset to hf dataset with our format

## Changes
- add pipeline for pile dataset to hf dataset

## Related Issues
Close #

## Checklist
- [ ] PR should be in the [Naming convention](../../docs/PR_NAMING.md)
- [ ] Assign yourself in to Assigneees
- [ ] Tag related issues
- [ ] Constants name should be ALL_CAPITAL, function name should be snake_case, and class name should be CamelCase
- [ ] complex function/algorithm should have [Docstring](https://peps.python.org/pep-0257/)
- [ ] 1 PR should not have more than 200 lines changes (Exception for test files). If more than that please open multiple PRs
- [ ] At least PR reviewer must come from the task's team (model, eval, data)

### CODE REVIEW

File: src/data/scripts/pile/convert_the_pile.py [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+import os
+import tqdm
+import json
+
+DATASET_NAME = "pile"
+NULL = "null"
+
+if __name__ == "__main__":
+    parser = ArgumentParser()
+    parser.add_argument("train_jsonl_uncompressed_directory")
```
Comment (new5558) [inline src/data/scripts/pile/convert_the_pile.py:14]: I just recently learn we can open `*.jsonl.zstd` without the need of uncompressing file.

Here is an example to do that
https://github.com/jzhang38/TinyLlama/blob/main/scripts/prepare_slimpajama.py

**Review by new5558 — CHANGES_REQUESTED** [review]

File: src/data/scripts/pile/convert_the_pile.py [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+    for jsonl_file in tqdm.tqdm(os.listdir(args.train_jsonl_uncompressed_directory)):
+        jsonl_file_path = os.path.join(
+            args.train_jsonl_uncompressed_directory, jsonl_file
+        )
+        with open(jsonl_file_path, "r", encoding="utf-8") as file:
+            i = 0
+            file_name = os.path.basename(jsonl_file).split(".")[0]
+            for line in file:
+                data = json.loads(line)
+                data_list["text"].append(data["text"])
```
Comment (boat1603) [inline src/data/scripts/pile/convert_the_pile.py:37]: ในโค้ดตรงนี้พี่มี concern OOM ในกรณีที่ data มีเยอะเกินไป พี่อยากแนะนำให้ใช้ `Dataset.from_generator` (https://huggingface.co/docs/datasets/create_dataset#from-local-files) ในการ handle ตรงนี้ ตัวอย่างเช่น

```python
def load_pile_dataset():
    for ...:
        yield {"text", "...", ...}
dataset = Dataset.from_generator(load_pile_dataset, num_proc=xxx)
```
ถ้าหากต้องการ pass parameter เข้าไปด้วยให้ใช้ [gen_kwargs](https://huggingface.co/docs/datasets/v2.14.5/en/package_reference/main_classes#datasets.Dataset.from_generator.example-2)

File: src/data/scripts/pile/convert_the_pile.py [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+                data_list["source"].append(DATASET_NAME)
+                data_list["source_id"].append(f"{file_name}_{i}")
+                data_list["create_date"].append(NULL)
+                data_list["update_date"].append(NULL)
+                data_list["meta"].append(data["meta"])
+                i += 1
+
+    dataset = Dataset.from_dict(data_list)
+    split = dataset.train_test_split(test_size=args.validation_size)
```
Comment (boat1603) [inline src/data/scripts/pile/convert_the_pile.py:46]: การใส่ `train_test_split` ควรใส่ seed เข้าไปด้วยไหม จะได้ reproduce ได้ง่าย

File: src/data/scripts/pile/convert_the_pile.py [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+    split["eval"] = split.pop("test")
+    split.save_to_disk(args.output_hf_directory)
```
Comment (boat1603) [inline src/data/scripts/pile/convert_the_pile.py:48]: เราควรใส่ `num_proc` ใน `save_to_disk` เพื่อ speed up การ save dataset

File: src/data/scripts/pile/convert_the_pile.py (hunk 1 above)
Comment (boss-chanon) [reply src/data/scripts/pile/convert_the_pile.py]: done

File: src/data/scripts/pile/convert_the_pile.py (hunk 2 above)
Comment (boss-chanon) [reply src/data/scripts/pile/convert_the_pile.py]: done

File: src/data/scripts/pile/convert_the_pile.py (hunk 3 above)
Comment (boss-chanon) [reply src/data/scripts/pile/convert_the_pile.py]: done

File: src/data/scripts/pile/convert_the_pile.py [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+        file_info = os.path.basename(jsonl_file).split(".")
+        # check zst compress
+        if file_info[-1] != "zst":
+            continue
+        file_name = file_info[0]
+        with zstd.open(open(jsonl_file_path, "rb"), "rt", encoding="utf-8") as file:
+            for line in file:
+                data = json.loads(line)
+                yield {
+                    "text": data["text"],
```
Comment (boat1603) [inline src/data/scripts/pile/convert_the_pile.py:26]: CONST

File: src/data/scripts/pile/convert_the_pile.py (hunk 4 above)
Comment (boss-chanon) [reply src/data/scripts/pile/convert_the_pile.py]: done

File: src/data/scripts/pile/convert_the_pile.py (hunk 5 above)
Comment (boss-chanon) [reply src/data/scripts/pile/convert_the_pile.py]: done

Comment (boat1603) [reply src/data/scripts/pile/convert_the_pile.py]: I mean all variables should be const

Comment (boss-chanon) [reply src/data/scripts/pile/convert_the_pile.py]: done

File: src/data/scripts/pile/convert_the_pile.py [hunk 6]
```diff
@@ -0,0 +1,59 @@
+from datasets import Dataset
+from argparse import ArgumentParser
+
+import zstandard as zstd
```
Comment (boat1603) [inline src/data/scripts/pile/convert_the_pile.py:4]: Should we add `zstandard` module to `pyproject.toml`?

Comment (boss-chanon) [reply src/data/scripts/pile/convert_the_pile.py:4]: zstandard is already in ```src/data/pyproject.toml```

**Review by boat1603 — APPROVED** [review]

**Review by new5558 — APPROVED** [review]
